src/_pytest/_code/code.py

```diff
@@ -721,11 +721,11 @@ def get_source(
     ) -> List[str]:
         """Return formatted and marked up source lines."""
         lines = []
-        if source is None or line_index >= len(source.lines):
+        if source is not None and line_index < 0:
+            line_index += len(source.lines)
```

Comment:
I think this is insufficient still -- for example if `len(source.lines)` is `4` and `line_index` is `-900` -- you could use `%` but it would probably be a good idea to figure out why these are negative at all (or refuse to produce source when the number is negative)

Reply:
That's why there's `line_index < 0` in the next condition :)
I'll try to investigate why this happens in the setuptools tests.